src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferencesSearchEngine.cs

```diff
@@ -66,92 +65,170 @@ public async Task FindReferencesAsync(ISymbol symbol, CancellationToken cancella
+            try
             {
-                foreach (var documentToFinderList in documentMap)
+                using var _1 = PooledHashSet<Document>.GetInstance(out var allDocuments);
+                foreach (var symbol in allSymbols)
                 {
-                    set.Clear();
-
-                    foreach (var tuple in documentToFinderList.Value)
-                        Debug.Assert(set.Add(tuple));
+                    foreach (var finder in _finders)
+                    {
+                        var documents = await finder.DetermineDocumentsToSearchAsync(
+                            symbol, project, _documents, _options, cancellationToken).ConfigureAwait(false);
+                        allDocuments.AddRange(documents);
+                    }
                 }
+
+                using var _2 = ArrayBuilder<Task>.GetInstance(out var tasks);
+                foreach (var document in allDocuments)
+                    tasks.Add(CreateWorkAsync(() => ProcessDocumentsAsync(document, allSymbols, cancellationToken), cancellationToken));
+
+                await Task.WhenAll(tasks).ConfigureAwait(false);
+            }
+            finally
+            {
+                await _progressTracker.ItemCompletedAsync(cancellationToken).ConfigureAwait(false);
             }
         }
 
-        private async ValueTask HandleLocationAsync(ISymbol symbol, ReferenceLocation location, CancellationToken cancellationToken)
+        private async Task ProcessDocumentsAsync(
+            Document document, ImmutableArray<ISymbol> symbols, CancellationToken cancellationToken)
         {
-            var group = await GetOrCreateSymbolGroupAsync(symbol, cancellationToken).ConfigureAwait(false);
-            await _progress.OnReferenceFoundAsync(group, symbol, location, cancellationToken).ConfigureAwait(false);
+            await _progress.OnFindInDocumentStartedAsync(document, cancellationToken).ConfigureAwait(false);
+
+            SemanticModel? model = null;
+            try
+            {
+                model = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+                // start cache for this semantic model
+                FindReferenceCache.Start(model);
+
+                foreach (var symbol in symbols)
+                    await ProcessDocumentAsync(document, model, symbol, cancellationToken).ConfigureAwait(false);
+            }
+            finally
+            {
+                FindReferenceCache.Stop(model);
+
+                await _progress.OnFindInDocumentCompletedAsync(document, cancellationToken).ConfigureAwait(false);
+            }
         }
 
-        private async ValueTask<SymbolGroup> GetOrCreateSymbolGroupAsync(ISymbol symbol, CancellationToken cancellationToken)
+        private static readonly Func<Document, ISymbol, string> s_logDocument = (d, s) =>
+            (d.Name != null && s.Name != null) ? string.Format("{0} - {1}", d.Name, s.Name) : string.Empty;
```

Comment:
document/symbol name would be PII, with @ryzngard's recent change to include telemetry for all these logblock events we should not include it.

Reply:
I'm definitely fine not including this. This was a move. But I'll remove it.